### ISSUE TITLE
TCP base SYN packet problem

### ISSUE THREAD

**Issue by LegacyOne** [0xcafed00d/tracetcp#5]
Hi it seems some sites like this:
tracetcp 46.19.169.70
are not replying to the TCP SYN that tracetcp sends out even with TTL 255 but the browser can connect fine. I have been able to track the problem to the TCP header length being 20 bytes which it would seem the site is filtering and dropping them.

If possible and I have tested this end can you make the TCP base SYN packet that tracetcp sends out with a header length of 24 bytes by including:
Kind: Maximum Segment Size (2)
Length: 4
MSS Value: 1460
with 00 00 padding  

Thanks

**Comment by 0xcafed00d** [comment]
Thanks for the report. Will investigate further.

**Comment by 0xcafed00d** [comment]
are you sure that the maximum segment size missing is the problem? I find that tracetcp connects to the address you supplied just fine: 

```
C:\bin>tracetcp 46.19.169.70 -h 200

Tracing route to 46.19.169.70 on port 80
Over a maximum of 30 hops.
200     Destination Reached in 33 ms. Connection established to 46.19.169.70
Trace Complete.
```

I have added the change you requested to a branch and have found that then tracetcp will not connect.

**Comment by LegacyOne** [comment]
Their must be a TCP SYN filter on a given route I'm on that your not if I do a test packet (using Colasoft Packet Builder) with maximum segment to make the TCP header length over 20 bytes with TTL 12 it works and I get a SYN ACK.

tracetcp 46.19.169.70 -h 200

Tracing route to 46.19.169.70 on port 80
Over a maximum of 30 hops.
200     \*       \*       \*       Request timed out.
201     \*       \*       \*       Request timed out.
202     \*       \*       \*       Request timed out.
203     \*       \*       \*       Request timed out.
204     \*       *

tracetcp 46.19.169.70 -m 255

Tracing route to 46.19.169.70 on port 80
Over a maximum of 255 hops.
1       7 ms    7 ms    7 ms    10.13.196.1
2       8 ms    9 ms    8 ms    62.252.175.1    [basl-core-2a-xe-210-0.network.virginmedia.net]
3       10 ms   9 ms    10 ms   81.96.0.154     [basl-core-2b-ae1-0.network.virginmedia.net]
4       \*       \*       \*       Request timed out.
5       13 ms   11 ms   17 ms   62.254.42.222   [nrth-bb-1c-ae1-0.network.virginmedia.net]
6       30 ms   34 ms   30 ms   62.254.42.178   [fran-ic-2-ae0-0.network.virginmedia.net]
7       31 ms   91 ms   29 ms   80.81.192.245   [xe-1-1-0.r1.bb-fo.fra1.vrsn.net]
8       32 ms   30 ms   30 ms   199.16.95.185   [ae1.666.r1.vidn-fo.fra1.vrsn.net]
9       \*       \*       \*       Request timed out.
10      \*       \*       \*       Request timed out.
11      \*       \*       \*       Request timed out.
12      \*       \*       \*       Request timed out.
13      \*       \*       \*       Request timed out.
14      \*       \*       \*       Request timed out.
15      \*       \*       \*       Request timed out.
16      \*       \*       \*       Request timed out.
17      \*       \*       \*       Request timed out.
18      \*       \*       \*       Request timed out.
19      \*       \*       \*       Request timed out.
20      \*       \*       \*       Request timed out.
21      \*       \*       \*       Request timed out.
22      \*       \*       \*       Request timed out.
23      \*       \*       \*       Request timed out.
24      \*       \*       \*       Request timed out.
25      \*       \*       \*       Request timed out.
26      \*       \*       \*       Request timed out.
27      \*       \*       \*       Request timed out.
28      \*       \*       \*       Request timed out.
29      \*       \*       \*       Request timed out.
30      \*       \*       \*       Request timed out.
31      \*       \*       \*       Request timed out.
32      \*       \*       \*       Request timed out.
33      \*       \*       \*       Request timed out.
34      \*       \*       \*       Request timed out.
35      \*       \*       \*       Request timed out.
36      \*       \*       \*       Request timed out.
37      \*       \*       \*       Request timed out.
38      \*       \*       \*       Request timed out.
39      \*       \*       \*       Request timed out.
40      \*       \*       \*       Request timed out.
41      \*       \*       \*       Request timed out.
42      \*       \*       \*       Request timed out.
43      \*       \*       \*       Request timed out.
44      \*       \*       \*       Request timed out.
45      \*       \*       \*       Request timed out.
46      \*       \*       \*       Request timed out.
47      *

**Comment by 0xcafed00d** [comment]
Could you capture the packet you are sending with wireshark, so I can compare it with the pa ket I am constructing.  Thanks.

**Comment by LegacyOne** [comment]
I can give you the hex of the test packet with the MAC, source IP and checksums missing in brackets here if thats ok:
From Wireshark.

[00 00 00 00 00 00] [00 00 00 00 00 00] 08 00 45 00
00 2c 09 d6 40 00 0c 06 [00 00] [00 00 00 00] 2e 13
a9 46 e4 f9 00 50 12 3e 25 b6 00 00 00 00 60 02
20 00 [00 00] 00 00 02 04 05 b4 00 00

**Comment by 0xcafed00d** [comment]
I am about to release v1.0.1 of tracetcp, which generates a SYN packet with the same tcp options that is sent when Chrome connects to a site, so if you can connect with chrome you should get a response with tracetcp. Let me know if this fixes the issue.

**Comment by LegacyOne** [comment]
Yes thats done it many thanks I was hoping for the SYN to be under 60 bytes for networking to know its just from tracetcp and not a SYN thats normally now seen but then someone might filter that too so with this you can't filter it of course it don't stop routers from not sending back time exceeded if they don't want too.

Only one smaller problem I've found which was their before the release is grc.com

tracetcp grc.com

Tracing route to 4.79.142.200 [grc.com] on port 80
Over a maximum of 30 hops.
1       7 ms    7 ms    7 ms    10.13.196.1
2       8 ms    20 ms   8 ms    62.252.175.1    [basl-core-2a-xe-210-0.network.virginmedia.net]
3       \*       \*       \*       Request timed out.
4       11 ms   9 ms    9 ms    62.254.42.198   [brnt-bb-1c-ae0-0.network.virginmedia.net]
5       12 ms   10 ms   10 ms   62.254.42.174   [m686-mp2.cvx1-b.lis.dial.ntli.net]
6       \*       \*       \*       Request timed out.
7       12 ms   12 ms   12 ms   84.116.136.94
8       18 ms   14 ms   14 ms   213.46.174.10
9       158 ms  154 ms  154 ms  4.69.160.10     [ae-21-3201.car1.Tustin1.Level3.net]
10      157 ms  153 ms  154 ms  4.69.160.10     [ae-21-3201.car1.Tustin1.Level3.net]
11      Destination Reached in 155 ms. Port closed on 4.79.142.200
Trace Complete.

It shows closed but if I do
tracetcp grc.com -h 12

Tracing route to 4.79.142.200 [grc.com] on port 80
Over a maximum of 30 hops.
12      Destination Reached in 153 ms. Connection established to 4.79.142.200
Trace Complete.

Its established easy fix is on reporting as closed send a SYN with a normal TTL to check.

**Comment by 0xcafed00d** [comment]
Glad that's working better now. There is not anything i can do with routers not replying with time exceeded messages, lots of sites don't respond to normal traceroutes either.
I'll investigate the grc.com issue, it seems that it is dropping the packet with a ttl=0, your suggested fix sounds like a reasonable solution

**Comment by 0xcafed00d** [comment]
I have just released v1.0.2. this should resolve the remaining issue. 

```
lmw@STARBLADE /C/development/github/tracetcp/Release/tracetcp_v1.0.2 (master)
$ ./tracetcp.exe grc.com

Tracing route to 4.79.142.200 [grc.com] on port 80
Over a maximum of 30 hops.
1       1 ms    1 ms    1 ms    192.168.1.1     [Wintermute]
2       8 ms    7 ms    8 ms    10.239.152.1
3       10 ms   7 ms    7 ms    62.252.175.137  [perr-core-2a-ae9-611.network.virginmedia.net]
4       *       *       *       Request timed out.
5       11 ms   10 ms   18 ms   62.254.42.242   [manc-bb-1c-ae1-0.network.virginmedia.net]
6       10 ms   11 ms   18 ms   77.67.65.141    [ae0.man11.ip4.gtt.net]
7       13 ms   28 ms   14 ms   141.136.106.253 [xe-3-0-5.lon10.ip4.gtt.net]
8       14 ms   14 ms   13 ms   4.68.111.25     [ae8.edge4.London.Level3.net]
9       311 ms  153 ms  272 ms  4.69.160.10     [ae-21-3201.car1.Tustin1.Level3.net]
10      199 ms  313 ms  226 ms  4.69.160.10     [ae-21-3201.car1.Tustin1.Level3.net]
11      Destination Reached in 147 ms. Connection established to 4.79.142.200
Trace Complete.
```

**Comment by LegacyOne** [comment]
Yep that's done it.
Thanks again
